### PR TITLE
Add default container volumes

### PR DESCRIPTION
**Pull request by pschorf**
## Changes proposed in this PR
- Adds support for default container images

## Why are we making these changes?
Supports adding default container volumes for all jobs.

### CODE REVIEW

File: scheduler/src/cook/mesos/task.clj [hunk 1]
```diff
@@ -77,10 +77,27 @@
             (assoc "EXECUTOR_PROGRESS_OUTPUT_FILE_ENV" "EXECUTOR_PROGRESS_OUTPUT_FILE_NAME"
                    "EXECUTOR_PROGRESS_OUTPUT_FILE_NAME" progress-output-file))))
 
+(defn merge-container-defaults
+  "Takes a job container specification and applies any defaults from the config"
+  [container]
+  (when container
+    (let [{:keys [volumes]} (config/container-defaults)
+          get-path (fn [volume] (or (:container-path volume)
+                                    (:host-path volume)))]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/task.clj]: Use destructuring?

File: scheduler/src/cook/mesos/task.clj [hunk 2]
```diff
@@ -77,10 +77,27 @@
...
+      (cond-> container
+              volumes
+              (update :volumes (fn [container-volumes]
+                                 (let [volumes-to-add (filter (fn [default-volume]
+                                                              (not-any? (fn [container-volume]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/task.clj]: Formatting seems off

File: scheduler/src/cook/mesos/task.clj [hunk 3]
```diff
@@ -77,10 +77,27 @@
...
+                                                                          (.startsWith (get-path default-volume)
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/task.clj]: Is `get-path` guaranteed to return a non-nil value?

File: scheduler/src/cook/mesos/task.clj [hunk 4]
```diff
@@ -77,10 +77,27 @@
...
+                                                                                       (get-path container-volume))) container-volumes))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/task.clj]: Nitpick, would have preferred the second argument to the `filter` (i.e. `container-volumes`) in a new line as the filter function spans over multiple lines.`

File: scheduler/src/cook/tools.clj [hunk 5]
```diff
@@ -180,7 +180,7 @@
 
 (defn job-ent->container
   "Take a job entity and return its container"
-  [db job-ent]
```
Comment (shamsimam) [inline scheduler/src/cook/tools.clj]: 👍

File: scheduler/src/cook/mesos/task.clj [hunk 6]
```diff
@@ -77,10 +77,27 @@
             (assoc "EXECUTOR_PROGRESS_OUTPUT_FILE_ENV" "EXECUTOR_PROGRESS_OUTPUT_FILE_NAME"
                    "EXECUTOR_PROGRESS_OUTPUT_FILE_NAME" progress-output-file))))
 
+(defn merge-container-defaults
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/task.clj]: Should this be named `merge-configured-volumes`?

File: scheduler/src/cook/mesos/task.clj [hunk 7]
```diff
@@ -77,10 +77,27 @@
...
+  "Takes a job container specification and applies any defaults from the config"
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/task.clj]: This function appears to only be merging the configured volumes?

File: scheduler/test/cook/test/mesos/task.clj [hunk 8]
```diff
@@ -875,3 +876,46 @@
...
+                                             :volumes [{:host-path "/mnt/app"
+                                                        :mode "rw"}]})))
+      (is (= {:docker {:image "foo"}
+              :volumes [{:host-path "/diff/a"
+                         :container-path "/c/a"}
+                        {:host-path "/mnt/app/data"
+                         :mode "r"}]}
+             (task/merge-container-defaults {:docker {:image "foo"}
+                                             :volumes [{:host-path "/diff/a"
+                                                        :container-path "/c/a"}]}))))))
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/task.clj]: Can we add a case for where `container-path` is `/c`?

File: integration/tests/cook/test_basic.py [hunk 9]
```diff
@@ -2838,3 +2838,68 @@ def test_cook_executor_reset_vars(self):
                 util.wait_for_instance(self.cook_url, uuid, status='success')
         finally:
             util.kill_jobs(self.cook_url, job_uuids, assert_response=False)
+
+
+    @unittest.skipUnless(util.has_docker_service() and util.docker_tests_enabled(), "Requires `docker inspect`")
+    def test_default_container_volumes(self):
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Can we write this test to have the command read and verify output of files mounted from the defaults and the user's mounted volumes to verify that mounts work as expected?

File: scheduler/test/cook/test/mesos/task.clj (hunk 8 above)
Comment (pschorf) [reply scheduler/test/cook/test/mesos/task.clj]: `/c/a/b`? I already have a test above for where the default is inside the user-provided volume. Added the other case.

**Comment by pschorf** [comment]
Pushed some integration test fixes.

**Pull request merged**